File: src/main/java/org/lanternpowered/server/scoreboard/LanternTeamBuilder.java
/*
 * Lantern
 *
 * Copyright (c) LanternPowered <https://www.lanternpowered.org>
 * Copyright (c) SpongePowered <https://www.spongepowered.org>
 * Copyright (c) contributors
 *
 * This work is licensed under the terms of the MIT License (MIT). For
 * a copy, see 'LICENSE.txt' or <https://opensource.org/licenses/MIT>.
 */
package org.lanternpowered.server.scoreboard;

import static com.google.common.base.Preconditions.checkArgument;
import static com.google.common.base.Preconditions.checkNotNull;
import static com.google.common.base.Preconditions.checkState;

import org.spongepowered.api.scoreboard.CollisionRule;
import org.spongepowered.api.scoreboard.CollisionRules;
import org.spongepowered.api.scoreboard.Team;
import org.spongepowered.api.scoreboard.Visibilities;
import org.spongepowered.api.scoreboard.Visibility;
import org.spongepowered.api.text.Text;
import org.spongepowered.api.text.format.TextColor;
import org.spongepowered.api.text.format.TextColors;

import java.util.HashSet;
import java.util.Set;

import org.checkerframework.checker.nullness.qual.Nullable;

public class LanternTeamBuilder implements Team.Builder {

    @Nullable private String name;
    @Nullable private Text displayName;
    private Text prefix;
    private Text suffix;
    private TextColor color;
    private boolean allowFriendlyFire;
    private boolean showFriendlyInvisibles;
    private Visibility nameTagVisibility;
    private Visibility deathMessageVisibility;
    private CollisionRule collisionRule;
    private Set<Text> members;

    public LanternTeamBuilder() {
        reset();
    }

    @Override
    public LanternTeamBuilder name(String name) {
        this.name = checkNotNull(name, "name");
        return this;
    }

    @Override
    public LanternTeamBuilder color(TextColor color) throws IllegalArgumentException {
        checkNotNull(color, "color");
        if (color == TextColors.RESET) {
            throw new IllegalArgumentException("Color cannot be TextColors.RESET!");
        }
        this.color = color;
        return this;
    }

    @Override
    public LanternTeamBuilder displayName(Text displayName) throws IllegalArgumentException {
        final int length = displayName.toPlain().length();
        checkArgument(length <= 32, "Display name is %s characters long! It must be at most 32.", length);
        this.displayName = displayName;
        return this;
    }

    @Override
    public LanternTeamBuilder prefix(Text prefix) {
        checkNotNull(prefix, "prefix");
        this.prefix = prefix;
        return this;
    }

    @Override
    public LanternTeamBuilder suffix(Text suffix) {
        checkNotNull(suffix, "suffix");
        this.suffix = suffix;
        return this;
    }

    @Override
    public LanternTeamBuilder allowFriendlyFire(boolean enabled) {
        this.allowFriendlyFire = enabled;
        return this;
    }

    @Override
    public LanternTeamBuilder canSeeFriendlyInvisibles(boolean enabled) {
        this.showFriendlyInvisibles = enabled;
        return this;
    }

    @Override
    public LanternTeamBuilder nameTagVisibility(Visibility visibility) {
        this.nameTagVisibility = checkNotNull(visibility, "visibility");
        return this;
    }

    @Override
    public LanternTeamBuilder deathTextVisibility(Visibility visibility) {
        this.deathMessageVisibility = checkNotNull(visibility, "visibility");
        return this;
    }

    @Override
    public LanternTeamBuilder collisionRule(CollisionRule rule) {
        this.collisionRule = checkNotNull(rule, "rule");
        return this;
    }

    @Override
    public LanternTeamBuilder members(Set<Text> members) {
        this.members = new HashSet<>(checkNotNull(members, "members"));
        return this;
    }

    @Override
    public LanternTeamBuilder from(Team value) {
        return name(value.getName())
                .displayName(value.getDisplayName())
                .color(value.getColor())
                .allowFriendlyFire(value.allowFriendlyFire())
                .canSeeFriendlyInvisibles(value.canSeeFriendlyInvisibles())
                .nameTagVisibility(value.getNameTagVisibility())
                .deathTextVisibility(value.getDeathMessageVisibility())
                .members(value.getMembers());
    }

    @Override
    public LanternTeamBuilder reset() {
        this.name = null;
        this.displayName = null;
        this.prefix = Text.of();
        this.suffix = Text.of();
        this.allowFriendlyFire = false;
        this.showFriendlyInvisibles = false;
        this.color(TextColors.NONE);
        this.nameTagVisibility(Visibilities.ALWAYS);
        this.deathTextVisibility(Visibilities.ALWAYS);
        this.collisionRule(CollisionRules.NEVER);
        this.members = new HashSet<>();
        return this;
    }

    @Override
    public Team build() throws IllegalStateException {
        checkState(this.name != null, "name must be set");

        Text displayName = this.displayName;
        if (displayName == null) {
            displayName = Text.of(this.name);
        }

        final LanternTeam team = new LanternTeam(this.name, this.color, displayName, this.prefix, this.suffix,
                this.allowFriendlyFire, this.showFriendlyInvisibles, this.nameTagVisibility, this.deathMessageVisibility,
                this.collisionRule);
        this.members.forEach(team::addMember);
        return team;
    }
}

File: src/main/kotlin/org/lanternpowered/server/scoreboard/LanternTeamBuilder.kt
/*
 * Lantern
 *
 * Copyright (c) LanternPowered <https://www.lanternpowered.org>
 * Copyright (c) SpongePowered <https://www.spongepowered.org>
 * Copyright (c) contributors
 *
 * This work is licensed under the terms of the MIT License (MIT). For
 * a copy, see 'LICENSE.txt' or <https://opensource.org/licenses/MIT>.
 */
package org.lanternpowered.server.scoreboard

import org.lanternpowered.api.scoreboard.ScoreboardTeam
import org.lanternpowered.api.scoreboard.ScoreboardTeamBuilder
import org.lanternpowered.api.text.Text
import org.lanternpowered.api.text.emptyText
import org.lanternpowered.api.text.format.NamedTextColor
import org.lanternpowered.api.text.textOf
import org.lanternpowered.api.text.toPlain
import org.lanternpowered.api.util.collections.toImmutableSet
import org.spongepowered.api.scoreboard.CollisionRule
import org.spongepowered.api.scoreboard.CollisionRules
import org.spongepowered.api.scoreboard.Visibilities
import org.spongepowered.api.scoreboard.Visibility

class LanternTeamBuilder : ScoreboardTeamBuilder {

    private var name: String? = null
    private var displayName: Text? = null
    private var prefix: Text = emptyText()
    private var suffix: Text = emptyText()
    private var color: NamedTextColor = NamedTextColor.WHITE
    private var allowFriendlyFire = false
    private var showFriendlyInvisibles = false
    private var nameTagVisibility: Visibility = Visibilities.ALWAYS.get()
    private var deathMessageVisibility: Visibility = Visibilities.ALWAYS.get()
    private var collisionRule: CollisionRule = CollisionRules.NEVER.get()
    private var members: Set<Text> = emptySet()

    override fun displayName(displayName: Text): LanternTeamBuilder = apply {
        val length = displayName.toPlain().length
        check(length <= 32) { "Display name is $length characters long! It must be at most 32." }
        this.displayName = displayName
    }

    override fun color(color: NamedTextColor): LanternTeamBuilder = apply { this.color = color }
    override fun name(name: String): LanternTeamBuilder = apply { this.name = name }
    override fun prefix(prefix: Text): LanternTeamBuilder = apply { this.prefix = prefix }
    override fun suffix(suffix: Text): LanternTeamBuilder = apply { this.suffix = suffix }
    override fun allowFriendlyFire(enabled: Boolean): LanternTeamBuilder = apply { this.allowFriendlyFire = enabled }
    override fun canSeeFriendlyInvisibles(enabled: Boolean): LanternTeamBuilder = apply { this.showFriendlyInvisibles = enabled }
    override fun nameTagVisibility(visibility: Visibility): LanternTeamBuilder = apply { this.nameTagVisibility = visibility }
    override fun deathTextVisibility(visibility: Visibility): LanternTeamBuilder = apply { this.deathMessageVisibility = visibility }
    override fun collisionRule(rule: CollisionRule): LanternTeamBuilder = apply { this.collisionRule = rule }
    override fun members(members: Set<Text>): LanternTeamBuilder = apply { this.members = members.toImmutableSet() }

    override fun from(value: ScoreboardTeam): LanternTeamBuilder = name(value.name)
            .displayName(value.displayName)
            .color(value.color)
            .allowFriendlyFire(value.allowFriendlyFire())
            .canSeeFriendlyInvisibles(value.canSeeFriendlyInvisibles())
            .nameTagVisibility(value.nameTagVisibility)
            .deathTextVisibility(value.deathMessageVisibility)
            .members(value.members)

    override fun reset(): LanternTeamBuilder = apply {
        this.name = null
        this.displayName = null
        this.prefix = emptyText()
        this.suffix = emptyText()
        this.allowFriendlyFire = false
        this.showFriendlyInvisibles = false
        this.color = NamedTextColor.WHITE
        this.nameTagVisibility = Visibilities.ALWAYS.get()
        this.deathMessageVisibility = Visibilities.ALWAYS.get()
        this.collisionRule = CollisionRules.NEVER.get()
        this.members = emptySet()
    }

    override fun build(): ScoreboardTeam {
        val name = checkNotNull(this.name) { "The name must be set" }
        val displayName = this.displayName ?: textOf(name)
        val team = LanternTeam(name, this.color, displayName, this.prefix, this.suffix,
                this.allowFriendlyFire, this.showFriendlyInvisibles, this.nameTagVisibility,
                this.deathMessageVisibility, this.collisionRule)
        this.members.forEach { member -> team.addMember(member) }
        return team
    }
}
